### ISSUE TITLE
REQUEST: GetCaps' structure values from retail hardware.

### ISSUE THREAD

**Issue by RadWolfie** [Cxbx-Reloaded/Cxbx-Reloaded#1132]
I need IDirectSound GetCaps' DSCAPS information in order to emulate it correctly before start working on DSound fixes part3.

2DBuffers = ?
3DBuffers = ?
BufferSGEs = should be 2047 (Is already emulated)
MemoryAllocated = should be 0 by default. (Is already emulated)

I have some information and theory for 2D and 3D Buffers value. However it is not enough info and really need hardware's info. Can someone please help me with this simple task? All it need to do is create directsound then call it. That's it.

BONUS: It would be nice to know what values all 4 variables above has after calling DirectSoundUseFullHRTF and DirectSoundUseLightHRTF separately to prove my theory.

**Comment by Fisherman166** [comment]
I took a shot at this last weekend. My first attempt was to use the dsound.h and x86 dsound.lib from the Windows 8.1 SDK. I was able to get it to compile and link with nxdk but I could not include include the xboxkrnl.h file due to name collision between Windows headers and xboxkrnl headers. Which ultimately killed the experiment because without xboxkrnl access there is no way to print out the information returned by the calls.

My second experiment was to stub out only the data structures I needed from dsound.h inside of main.c. and then link in the x86 dsound.lib file, which would allow me to use xboxkrnl functions. Again I was able to get this to compile under nxdk but my Xbox hung when executing DirectSoundCreate.

I'm stumped at this point. I'll include the code to both of my experiments so someone else that knows what they are doing can maybe make this happen.

[audo_query.zip](https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/files/2020012/audo_query.zip)

**Comment by RadWolfie** [comment]
Windows' DSound library is not the same as Xbox's DSound. You have to use Xbox's DSound library in order for xbox's hardware to work correctly.

Nxdk's xboxkrnl functions might not be accurate or not compatible enough. This is my guess. Though, I don't think this is the cause of the problem. I think you are linking to Windows' DSound library instead of Xbox's library.

**Comment by Fisherman166** [comment]
That makes perfect sense. The only problem is that the nxdk does not have any audio libraries as far as I can tell. Which is why I tried the windows library since I didn't any have any other ideas.

**Comment by PatrickvL** [comment]
If I understand correctly, the whole idea is to get information about the Xbox direct sound capabilities record, which can only be retrieved by means of an XDK build software title that calls GetCaps for dsound

**Comment by RadWolfie** [comment]
I'll make it simple.
```
Nxdk (community open source 2nd version) =/= XDK (Microsoft's xbox development kit)
```
If you don't have XDK source files, then don't waste your time trying to get inaccessible xbox DSound library to work. Unless you do, then use the xbox directory inside xdk which contain the xbox's version of DSound.

**Comment by x1nixmzeng** [comment]
Another approach @Fisherman166 would be to a game which calls this method and patch the XBE to dump information to a local file.

**Comment by LukeUsher** [comment]
I haven't done testing on real hardware yet (I will later :P), but from disassembling various XBEs I found the default value that these fields are initialized to:

3D Buffers = 64
2D Buffers = 192

Adding up to a total of 256 buffers.

Obviously these values are the available buffers, so they should decrement/increment as buffers are created and freed.

**Comment by RadWolfie** [comment]
Yeah, my investigation lead to same result as you have base on the documentation. For some reason, it still wants to create 2047 classes of DirectSoundStream. Even if I return out of memory error.

I think two of test cases made my code not to work correctly was Mafia and Burnout.

Note for others to remember. Some titles want to create 2047 classes of Buffer, others for Stream class. Then it stop there. So, that's outside of 256 buffers limit.

**Comment by RadWolfie** [comment]
**UPDATE**

Turns out I am able to use both trampoline and lle audio experiment to figure out where the initialized values came from. It is from software's end instead of hardware directly.

**Comment by RadWolfie** [comment]
**ANOTHER UPDATE**

I have new idea to resolve this issue. However, the requirement is partial stubbed LLE APU in order to allow to remove GetCaps function patch. Plus trampoline method to forward handle the allocate/deallocate the buffer and stream classes. If trampoline return a fail status, do nothing plus send the status back to the title.
